terraform/jenkins/jenkins_controller.tf

```diff
@@ -0,0 +1,148 @@
+# SPDX-FileCopyrightText: 2023 Technology Innovation Institute (TII)
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# Build the Jenkins controller image
+module "jenkins_controller_image" {
+  source = "../../tf-modules/azurerm-nix-vm-image"
+
+  nix_attrpath   = "outputs.nixosConfigurations.jenkins-controller.config.system.build.azureImage"
+  nix_entrypoint = "${path.module}/../.."
+
+
+  name                = "jenkins-controller"
+  resource_group_name = azurerm_resource_group.default.name
+  location            = azurerm_resource_group.default.location
+
+  storage_account_name   = azurerm_storage_account.vm_images.name
+  storage_container_name = azurerm_storage_container.vm_images.name
+}
+
+# Create a machine using this image
+module "jenkins_controller_vm" {
+  source = "../../tf-modules/azurerm-linux-vm"
+
+  resource_group_name = azurerm_resource_group.default.name
+  location            = azurerm_resource_group.default.location
+
+  virtual_machine_name         = "ghaf-jenkins-controller"
+  virtual_machine_size         = "Standard_D1_v2"
+  virtual_machine_source_image = module.jenkins_controller_image.image_id
+
+  virtual_machine_custom_data = join("\n", ["#cloud-config", yamlencode({
+    users = [
+      for user in toset(["bmg", "flokli", "hrosten"]) : {
+        name                = user
+        sudo                = "ALL=(ALL) NOPASSWD:ALL"
+        ssh_authorized_keys = local.ssh_keys[user]
+      }
+    ]
+    write_files = [
+      # See corresponding EnvironmentFile= directives in services
+      {
+        content = "KEY_VAULT_NAME=${azurerm_key_vault.ssh_remote_build.name}\nSECRET_NAME=${azurerm_key_vault_secret.ssh_remote_build.name}",
+        "path"  = "/var/lib/fetch-build-ssh-key/env"
```

Comment:
so, ssh_remote_build.name is stored as a secret in keyvault?

Reply:
This configures the EnvironmentFile= that the `fetch-build-ssh-key` systemd service sees.

It sets `KEY_VAULT_NAME` to the name of the vault, and `SECRET_NAME` to the name of the secret in that vault.

Both of them are interpolated through terraform from the attributes in the corresponding resources, which properly models the dependency graph, and ensures things are created in the right order.

Rendered on disk, `/var/lib/fetch-build-ssh-key/env` currently has the following contents:

```
KEY_VAULT_NAME=ghaf-ssh-remote-build
SECRET_NAME=remote-build-ssh-private-key
```

and no, none of these values are secret, they can also be inferred by staring at the terraform code ;-)